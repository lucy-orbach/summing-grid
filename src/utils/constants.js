export const app ={
	FIELDS: 3,
};