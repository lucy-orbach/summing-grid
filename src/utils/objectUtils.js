export default class ObjectUtils {
	static iterateOverNumber = (num, func) => {
		for ( let i of Array(num).keys() ) {
			func(i);
		}
	};
}