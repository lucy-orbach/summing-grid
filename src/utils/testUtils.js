import 'setupTests';
import { shallow, find, simulate } from 'enzyme';

export default class TestUtils {
	static updateInput = ( wrapper, id, value ) => {
		let selector =  `[data-test="input-${id}"]`;
		const input = wrapper.find(selector);
		input.simulate( 'change', { target: { name: id, value: value }} );

		return wrapper.find( selector );
	};


}