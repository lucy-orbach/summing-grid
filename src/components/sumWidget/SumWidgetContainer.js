import React, { Component } from 'react';
import WidgetDynamicTitle from 'components/sumWidget/WidgetDynamicTitle';
import NumInput from 'components/formElements/NumInput';
import * as CONSTANTS from 'utils/constants';
import styles from 'components/sumWidget/SumWidgetContainer.module.css';

let { FIELDS } = CONSTANTS.app;

export default class SumWidgetContainer extends Component {

	handleInputChange = () => {
		console.log('handleInputchange');
	};

	renderInputFields = () => {
		let inputFields = [];
		for ( let i of Array(FIELDS).keys() ) {
			inputFields.push(
				<li key={i}>
					<NumInput value={i} error={""} onChange={this.handleInputChange} />
				</li>
			);
		}
		return inputFields;
	};

	render() {
		console.log(this.state);
		return (
			<section className={styles.container}>
				<WidgetDynamicTitle title={"hi"} />
				<ul className={styles.list}>
					{this.renderInputFields()}
					<li>Total</li>
				</ul>
			</section>
		);
	}
}