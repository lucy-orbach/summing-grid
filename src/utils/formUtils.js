import { ValueKeys } from 'utils/constants';

export default class FormUtils {
	static clearInputValue = value => value.split(/[ ,]+/).join('');

	static formatNumericValue = value => value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");

	static validateNumber = value => !isNaN(value);

	// format number to nearest value
	static formatToNearestValue = value => {
		for ( let key of Object.keys(ValueKeys) ) {
			let exp = ValueKeys[key];
			if ( value >= exp) {
				if (key === 'BIG') return FormUtils.getVeryBigNumber(value);
				return `${(value/exp).toPrecision(3)} ${key}`;
			}
		}

		return value.toPrecision(3);
	};

	// returns a scientific notation for large number
	static getVeryBigNumber = value => {
		let digits = value.toString().split('');
		let rest = digits.splice(3);
		let newValue = (parseInt(digits.join(''))/100).toPrecision(3);
		let exp = rest.length + 1;

		return `${newValue} x 10e+${exp}`;
	};
};