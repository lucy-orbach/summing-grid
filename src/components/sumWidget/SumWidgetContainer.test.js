import React from 'react';
import 'setupTests';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SumWidgetContainer from 'components/sumWidget/SumWidgetContainer';
import TestUtils from 'utils/TestUtils';
import * as CONSTANTS from 'utils/constants';
//
let { FIELDS } = CONSTANTS.appConstants;

describe('<SumWidgetContainer />', () => {

	it('should render SumWidgetContainer component ', () => {
		const wrapper = shallow(<SumWidgetContainer />);
		expect( wrapper.find('[data-test="sum_widget_container"]').exists()).toBe(true);
	});

	it('Fields Number should be from 2 - 100', () => {
		let wrapper = shallow(<SumWidgetContainer fieldsNumber={100}/>);
		expect(wrapper.find('[data-test="total"]').exists()).toBe(true);

		wrapper = shallow(<SumWidgetContainer fieldsNumber={1} />);
		expect(wrapper.find('[data-test="total"]').exists()).toBe(false);

		 wrapper = shallow(<SumWidgetContainer fieldsNumber={101} />);
			expect(wrapper.find('[data-test="total"]').exists()).toBe(false);
	});


	it('should add correct amount of fields', () => {
		const wrapper = shallow(<SumWidgetContainer />);
		expect(Object.keys(wrapper.state().fields).length).toBe(FIELDS);
	});


	it('should render correct amount of Inputs',() => {
		const wrapper = shallow(<SumWidgetContainer />);
		expect( wrapper.find('li').length).toBe(4);
	});


	it('input value should clear when current value is 0', () => {
		const wrapper = shallow(<SumWidgetContainer />);
		const ele = wrapper.find('[data-test="input-0"]');
		ele.simulate('click', { target:{ name: 0, value: "0" }});
		expect( wrapper.find('[data-test="input-0"]').props().value).toBe('');

		ele.simulate('click', { target:{ name: 0, value: "30" }});
		expect( wrapper.find('[data-test="input-0"]').props().value).toBe('30');
	});


	it('user should be able to enter value on input', () => {
		const wrapper = shallow(<SumWidgetContainer />);
		const updatedInput = TestUtils.updateInput(wrapper,1,'30');
		expect( updatedInput.props().value).toBe('30');
	});


  it('on input change, value should remove format and set on state', () => {
	  const wrapper = shallow(<SumWidgetContainer />);
	  // remove comas
	  let updatedInput = TestUtils.updateInput(wrapper,1,'3,000');
	  expect( wrapper.state().fields[1].value).toBe('3000');
	  // remove  whitespaces
	   updatedInput = TestUtils.updateInput(wrapper,1,' 3,00 0');
	  expect( wrapper.state().fields[1].value).toBe('3000');
  });


  it('if not a number => should display value entered, add error to field and display error text', () => {
	  const wrapper = shallow(<SumWidgetContainer />);
	  // remove comas
	  let updatedInput = TestUtils.updateInput(wrapper,1,'3r');
	  expect( wrapper.state().fields[1].value).toBe('3r');
	  expect( wrapper.state().fields[1].error).toBe(true);
  });


  it('should sum valid values', () => {
	  const wrapper = shallow(<SumWidgetContainer />);
	  expect( wrapper.state().sum).toBe(0);
	  let updateInput0 = TestUtils.updateInput(wrapper,0,'3');
	  expect( wrapper.state().sum).toBe(3);
	  let updateInput1 = TestUtils.updateInput(wrapper,1,'5x');
	  expect( wrapper.state().sum).toBe(3);
	   updateInput1 = TestUtils.updateInput(wrapper,1,'5');
	  expect( wrapper.state().sum).toBe(8);
	  let updateInput2 = TestUtils.updateInput(wrapper,2,'');
	  expect( wrapper.state().sum).toBe(8);
	  updateInput2 = TestUtils.updateInput(wrapper,2,'2');
	  expect( wrapper.state().sum).toBe(10);
  });


  it('should display formatted total', () => {
	  const wrapper = shallow(<SumWidgetContainer />);

	  expect( wrapper.find('[data-test="total"]').text()).toBe("0.00");
	  let updatedInput = TestUtils.updateInput(wrapper,1,'1.567');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.57');
	  updatedInput = TestUtils.updateInput(wrapper,1,'15.67');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('15.7');
	  updatedInput = TestUtils.updateInput(wrapper,1,'156.7');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('157');
	  updatedInput = TestUtils.updateInput(wrapper,1,'1567');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.57 K');
	  updatedInput = TestUtils.updateInput(wrapper,1,'1567000');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.57 M');
	  updatedInput = TestUtils.updateInput(wrapper,1,'1567000000');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.57 B');
	  updatedInput = TestUtils.updateInput(wrapper,1,'1567000000000');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.57 T');
	  updatedInput = TestUtils.updateInput(wrapper,1,'1567000000000000');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.57 Q');
	  updatedInput = TestUtils.updateInput(wrapper,1,'1567000000000000000');
	  expect( wrapper.find('[data-test="total"]').text()).toBe('1.56 x 10e+17');
  });


	it('matches the snapshot', () => {
		const tree = shallow(<SumWidgetContainer />);
		expect(toJson(tree)).toMatchSnapshot()
	});

});