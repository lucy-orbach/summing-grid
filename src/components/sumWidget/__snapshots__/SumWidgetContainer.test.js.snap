// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SumWidgetContainer /> matches the snapshot 1`] = `
<section
  className="container"
  data-test="sum_widget_container"
>
  <React.Memo
    dataTest="widget_title"
    hasError={false}
    title="Hi! Let's sum 3 values :)"
  />
  <ul
    className="list"
  >
    <li
      className="cell"
      key="0"
    >
      <React.Memo
        data-test="input-0"
        error={null}
        name={0}
        onChange={[Function]}
        onClick={[Function]}
        value={0}
      />
    </li>
    <li
      className="cell"
      key="1"
    >
      <React.Memo
        data-test="input-1"
        error={null}
        name={1}
        onChange={[Function]}
        onClick={[Function]}
        value={0}
      />
    </li>
    <li
      className="cell"
      key="2"
    >
      <React.Memo
        data-test="input-2"
        error={null}
        name={2}
        onChange={[Function]}
        onClick={[Function]}
        value={0}
      />
    </li>
    <li
      className="cell"
      data-test="total"
      key="total"
    >
      0.00
    </li>
  </ul>
</section>
`;
